doc/developer/guide-changes.md

```diff
@@ -0,0 +1,790 @@
+
+Fixing a bug in a dependency is much harder than fixing a bug in core
+Materialize code. You'll have to check out the repository, spin up on the
+codebase, fix the bug, use Cargo's somewhat buggy [patch feature][cargo-patch]
+to verify the fix in Materialize, then send a PR to upstream. If upstream is
+responsive, then you just need to wait for a new version with the fix on
+crates.io. If upstream is not responsive, you'll need to fork the dependency,
+then keep that fork up-to-date if upstream ever becomes active again. Each step
+is a small papercut, but they add up.
+
+The best case is when a dependency is as actively maintained as Materialize.
+Dependencies that fall into this category include [Tokio], [rust-postgres],
+and [reqwest]. These dependencies have largely had no bugs at all. When they
+have had bugs, the upstream maintainers have been very quick to fix them.
+
+Use caution with dependencies that are not actively maintained. Materialize
+often winds up taking over maintainership of these dependencies. This can make
+sense for dependencies within our core competencies, like [rust-rdkafka]
+and [sqlparser], but it's a real drag for peripherals, like [rust-prometheus].
+
+### Documentation
+
+Documentation means at least two things in the repository: code documentation
+and user-facing documentation. Both of these are important, but have different
+roles. While the act of documentation may feel less important than the production
+of code, the enduring value of the code is greatly diminished without clear
+documentation. Clear documentation is a required component of a code contribution.
+
+Code documentation is important to explain the intended behavior of your code
+without requiring a reader to extract this from your code directly. It serves
+as a contract between your implementation and the users of your code. The value
+of clear documentation lies in the extent to which it can reduce the time others
+must spend inspecting your code to determine its appropriateness. Documentation
+also provides for continuity as others may need to work with and take ownership
+of code that you have written.
+
+A PR without appropriate code documentation may reasonably be rejected. Imagine
+that the reviewer would first like to read a precis of your new code, and only
+the read the code itself. This approximates the experience of most people new
+to your new code.
+
+User documentation is also important, but may be outside the tradecraft that
+you are trained for. If you are creating a new user-facing feature, please
+ensure that appropriate documentation is produced, either as part of the pull
+request if that is possible, or scheduled appropriately as the feature is
+released.
+
+### Testing
+
+With perfect testing, introducing a bug in any line of any function in the
+codebase would cause at least one test to fail. Meeting this standard would
+unacceptably impair velocity, but it's a useful framing for what ideal test
+coverage means.
```

Comment:
Also with perfect testing *not* introducing a bug should only ever require *adding* tests.

Reply:
I'm a bit confused by this phrasing. If you have perfect testing, then by definition you don't need to add any tests?